### PR TITLE
dockerregistry tests

### PR DESCRIPTION
**Pull request by legionus**
Add more tests for `dockerregistry`:

```
ok github.com/openshift/origin/pkg/dockerregistry/server 2.046s coverage: 54.4% of statements
```

### CODE REVIEW

**Comment by legionus** [comment]
[test]

**Comment by legionus** [comment]
caused by #12452 
re[test]

**Comment by ncdc** [comment]
@legionus you need to update your code post rebase. Some of the kube test packages changed.

**Comment by legionus** [comment]
[test]

**Comment by legionus** [comment]
@miminar please review

**Comment by legionus** [comment]
I need `GetFakeImageStreamImageGetHandler` to create tests for https://github.com/openshift/origin/pull/12504

**Comment by legionus** [comment]
[test]

**Comment by aweiteka** [comment]
[Discover the master API URL:PORT given only a registry endpoint](https://trello.com/c/NsI8oO9S/811-discover-the-master-api-url-port-given-only-a-registry-endpoint)

File: pkg/dockerregistry/testutil/manifests.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+	_, payload, err := manifest.Payload()
+	if err != nil {
+		return "", nil, err
+	}
+
+	return string(payload), manifest, nil
+}
+
+func MakeLayers(layerCount int) ([]distribution.Descriptor, []LayerPayload, error) {
```
Comment (miminar) [inline pkg/dockerregistry/testutil/manifests.go]: `MakeRandomLayers`

File: pkg/dockerregistry/testutil/manifests.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+		}
+
+		layers = append(layers, distribution.Descriptor{Digest: dgst, Size: int64(len(content))})
+		payloads = append(payloads, LayerPayload(content))
+	}
+
+	return layers, payloads, nil
+}
+
+func MakeConfig() (ConfigPayload, error) {
```
Comment (miminar) [inline pkg/dockerregistry/testutil/manifests.go]: `MakeManifestConfig`

File: pkg/dockerregistry/testutil/manifests.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+func CreateRandomManifest(schemaVersion ManifestSchemaVesion, layerCount int) (string, distribution.Manifest, *Payload, error) {
+	var (
+		rawManifest string
+		manifest    distribution.Manifest
+		err         error
+	)
+
+	layersDescs, layerPayloads, err := MakeLayers(layerCount)
+	if err != nil {
+		return "", nil, nil, fmt.Errorf("unable generate layers: %v", err)
```
Comment (miminar) [inline pkg/dockerregistry/testutil/manifests.go]: cannot generate

File: pkg/dockerregistry/testutil/manifests.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+func MakeConfig() (ConfigPayload, error) {
+	cfg := imageapi.DockerImageConfig{}
+
+	jsonBytes, err := json.Marshal(&cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return jsonBytes, nil
```
Comment (miminar) [inline pkg/dockerregistry/testutil/manifests.go]: Could this return config descriptor as well?

File: pkg/dockerregistry/testutil/util.go [hunk 5]
```diff
@@ -266,6 +266,45 @@ func TestNewImageStreamObject(namespace, name, tag, imageName, dockerImageRefere
...
+// with wake os client returning one of given imagestream image objects if found.
+func GetFakeImageStreamImageGetHandler(t *testing.T, ism *imageapi.ImageStream, iss ...imageapi.Image) core.ReactionFunc {
+	return func(action core.Action) (handled bool, ret runtime.Object, err error) {
+		switch a := action.(type) {
+		case core.GetAction:
+			for _, is := range iss {
+				name, imageID, err := imageapi.ParseImageStreamImageName(a.GetName())
+				if err != nil {
+					return true, nil, err
+				}
```
Comment (miminar) [inline pkg/dockerregistry/testutil/util.go]: Move this before the switch.

File: pkg/dockerregistry/testutil/util.go [hunk 6]
```diff
@@ -266,6 +266,45 @@ func TestNewImageStreamObject(namespace, name, tag, imageName, dockerImageRefere
 	}
 }
 
+// GetFakeImageStreamImageGetHandler returns a reaction function for use
+// with wake os client returning one of given imagestream image objects if found.
+func GetFakeImageStreamImageGetHandler(t *testing.T, ism *imageapi.ImageStream, iss ...imageapi.Image) core.ReactionFunc {
```
Comment (miminar) [inline pkg/dockerregistry/testutil/util.go]: `s/ism/iss/`
`s/iss/imgs/`

File: pkg/dockerregistry/testutil/util.go [hunk 7]
```diff
@@ -266,6 +266,45 @@ func TestNewImageStreamObject(namespace, name, tag, imageName, dockerImageRefere
...
+
+				if imageID != is.Name {
+					continue
+				}
+
+				t.Logf("imagestreamimage get handler: returning image %s", is.Name)
+
+				isi := imageapi.ImageStreamImage{
+					ObjectMeta: kapi.ObjectMeta{
+						Namespace:         is.Namespace,
```
Comment (miminar) [inline pkg/dockerregistry/testutil/util.go]: The `Namespace` needs to come from image stream, not image.

File: pkg/dockerregistry/testutil/util.go [hunk 8]
```diff
@@ -266,6 +266,45 @@ func TestNewImageStreamObject(namespace, name, tag, imageName, dockerImageRefere
...
 }
 
+// GetFakeImageStreamImageGetHandler returns a reaction function for use
+// with wake os client returning one of given imagestream image objects if found.
+func GetFakeImageStreamImageGetHandler(t *testing.T, ism *imageapi.ImageStream, iss ...imageapi.Image) core.ReactionFunc {
+	return func(action core.Action) (handled bool, ret runtime.Object, err error) {
+		switch a := action.(type) {
+		case core.GetAction:
+			for _, is := range iss {
+				name, imageID, err := imageapi.ParseImageStreamImageName(a.GetName())
```
Comment (miminar) [inline pkg/dockerregistry/testutil/util.go]: `name` should be compared against image stream name where the image is tagged.

File: pkg/dockerregistry/server/pullthroughmanifestservice_test.go [hunk 9]
```diff
@@ -156,11 +167,18 @@ func TestPullthroughManifests(t *testing.T) {
 			if tc.expectedNotFoundError && err == distribution.ErrBlobUnknown {
 				break
 			}
+			if e, ok := err.(*kerrors.StatusError); ok {
+				if tc.expectedNotFoundError && e.ErrStatus.Code == http.StatusNotFound {
```
Comment (legionus) [inline pkg/dockerregistry/server/pullthroughmanifestservice_test.go]: @miminar But this error is [coming back now](https://github.com/openshift/origin/blob/master/pkg/dockerregistry/server/pullthroughmanifestservice.go#L42). These tests I checked with `go test -cover`. More examples:

https://github.com/openshift/origin/blob/master/pkg/dockerregistry/server/manifestservice.go#L195
https://github.com/openshift/origin/blob/master/pkg/dockerregistry/server/manifestservice.go#L204
https://github.com/openshift/origin/blob/master/pkg/dockerregistry/server/manifestservice.go#L214
https://github.com/openshift/origin/blob/master/pkg/dockerregistry/server/tagservice.go#L159

Comment (miminar) [inline pkg/dockerregistry/server/pullthroughmanifestservice_test.go]: Yes, but it's wrong. StatusError will be converted to 500 unknown error. We should propagate the information to the client where possible.

If you don't want to touch that in this PR, please add at least a TODO here to rewrite kube errors to distribution errors in the middleware.

File: pkg/dockerregistry/server/tagservice_test.go [hunk 10]
```diff
@@ -0,0 +1,976 @@
...
+	defer func() {
+		// set it back once this test finishes to make other unit tests working again
+		DefaultRegistryClient = backupRegistryClient
+	}()
+
+	ctx := context.Background()
+	serverURL, _ := url.Parse("docker.io/centos")
+
+	testImage := createTestImage(t, client, serverURL, namespace, repo)
+	createTestImageStream(t, client, testImage, namespace, repo, tag)
```
Comment (legionus) [inline pkg/dockerregistry/server/tagservice_test.go]: @miminar Not yet.

**Comment by legionus** [comment]
@miminar please review again

File: pkg/dockerregistry/server/tagservice_test.go [hunk 11]
```diff
@@ -0,0 +1,922 @@
...
+			tagName:      tag + "-new",
+			tagValue:     distribution.Descriptor{Digest: digest.Digest(testImage.Name)},
+			pullthrough:  true,
+			imageManaged: true,
+		},
+		{
+			title:         "create tag on unmanaged image without pullthrough",
+			tagName:       tag + "-new",
+			tagValue:      distribution.Descriptor{Digest: digest.Digest(testImage.Name)},
+			expectedError: true,
```
Comment (miminar) [inline pkg/dockerregistry/server/tagservice_test.go]: Could the error be more specific? This cases should be properly defined.

File: pkg/dockerregistry/server/tagservice_test.go [hunk 12]
```diff
@@ -0,0 +1,922 @@
...
+		case nil:
+			if tc.expectedError || tc.expectedNotFoundError {
+				t.Fatalf("[%s] unexpected successful response", tc.title)
+			}
+		default:
+			if tc.expectedError {
+				break
+			}
+			t.Fatalf("[%s] unexpected error: %#+v", tc.title, err)
+		}
```
Comment (miminar) [inline pkg/dockerregistry/server/tagservice_test.go]: Can you try to get the tag again?

File: pkg/dockerregistry/server/tagservice_test.go [hunk 13]
```diff
@@ -0,0 +1,976 @@
...
+	name reference.Named
+}
+
+func (r *testRepository) Named() reference.Named {
+	return r.name
+}
+
+type testTagService struct {
+	data  map[string]distribution.Descriptor
+	calls map[string]int
```
Comment (miminar) [inline pkg/dockerregistry/server/tagservice_test.go]: bump :)

File: pkg/dockerregistry/server/tagservice_test.go [hunk 14]
```diff
@@ -0,0 +1,976 @@
...
+				ctx:              ctx,
+				namespace:        namespace,
+				name:             repo,
+				pullthrough:      tc.pullthrough,
+				cachedLayers:     cachedLayers,
+				registryOSClient: client,
+			},
+		}
+
+		err = ts.Tag(ctx, tc.tagName, tc.tagValue)
```
Comment (miminar) [inline pkg/dockerregistry/server/tagservice_test.go]: Still applies.

**Comment by legionus** [comment]
@miminar I will change it in following PR if you don't mind.

**Comment by legionus** [comment]
@mfojtik merge?

**Comment by mfojtik** [comment]
[merge]

**Comment by legionus** [comment]
@mfojtik flake #12236 again. merge ?

**Comment by mfojtik** [comment]
[merge]

**Pull request merged**